src/dwarf_parser.cpp

```diff
@@ -247,18 +247,49 @@ void Dwarf::resolve_fields(const SizedType &type)
     return;
 
   auto type_dbg = target_.FindFirstType(type_name.c_str());
-  if (!type_dbg)
+  if (!type_dbg.IsValid())
+    return;
+
+  resolve_fields(std::move(str), std::move(type_dbg));
+}
+
+void Dwarf::resolve_fields(std::shared_ptr<Struct> str, lldb::SBType type)
+{
+  if (!str || str->HasFields())
     return;
 
-  for (uint32_t i = 0; i < type_dbg.GetNumberOfFields(); i++) {
-    auto field = type_dbg.GetFieldAtIndex(i);
-    auto field_type = get_stype(field.GetType());
-    str->AddField(field.GetName() ?: "",
-                  get_stype(field.GetType()),
-                  field.GetOffsetInBytes(),
-                  resolve_bitfield(field),
-                  false);
+  if (!type.IsValid())
+    return;
+
+  std::queue<Subobject> subobjects{ std::deque{
+      Subobject{ std::move(type), 0 } } };
+  while (!subobjects.empty()) {
+    auto &subobject = subobjects.front();
+
+    // Collect the fields into str, duplicates will be ignored.
+    for (uint32_t i = 0; i < subobject.type.GetNumberOfFields(); i++) {
+      auto field = subobject.type.GetFieldAtIndex(i);
+      str->AddField(field.GetName() ?: "",
+                    get_stype(field.GetType()),
+                    subobject.offset + field.GetOffsetInBytes(),
+                    resolve_bitfield(field),
+                    false);
+    }
+
+    // Queue the bases for further processing.
+    for (uint32_t i = 0; i < subobject.type.GetNumberOfDirectBaseClasses();
+         i++) {
+      auto base = subobject.type.GetDirectBaseClassAtIndex(i);
+      subobjects.push(Subobject{ base.GetType(),
+                                 subobject.offset + base.GetOffsetInBytes() });
+    }
+
+    subobjects.pop();
   }
+
+  std::sort(str->fields.begin(),
+            str->fields.end(),
+            [](const Field &a, const Field &b) { return a.offset < b.offset; });
```

Comment:
Don't forget to test this please!

We need tests for any behaviour that we want to ensure is preserved across potential future refactors or rewrites